Add Batch Rule for aten::narrow.Tensor

isdanni opened the pull request:
Summary:
Add batch support for `aten::narrow.Tensor`.

Similar logic to `narrow_copy`

Fix: #99558 

Test Plan: Please SEE contbuild & OSS CI.

Differential Revision: D45197155

zou3519 commented:
Test failures look real

zou3519 commented:
So, the error is saying that narrow.Tensor is registered as a CompositeImplicitAutograd operator, and we do not register batching rules for CompositeImplicitAutograd operations.

The high-level approach to the fix is to make narrow.Tensor not CompositeImplicitAutograd. narrow.Tensor is not supposed to be CompositeImplicitAutograd, because its implementation calls .data_ptr(). That is, we need to define a backward formula for it, and then mark it as CompositeExplicitAutograd.

kshitij12345 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4012,6 +4012,14 @@
   dispatch:
     CompositeImplicitAutograd: narrow_tensor_symint
```
You will have to update this to `CompositeExplicitAutograd` and add entry in `derivatives.yaml`.

isdanni replied:
> You will have to update this to `CompositeExplicitAutograd` and add entry in `derivatives.yaml`.

Thank you so much!!!  A follow up question: I got this `AssertionError: narrow_backward appears to be a new operator that aliases its inputs. The codegen expects you to add a corresponding operator to native_functions.yaml: narrow_backward_copy. See Note [view_copy NativeFunctions] for details.` in a [failed job](https://github.com/pytorch/pytorch/actions/runs/4833926807/jobs/8614534323).

Do I need to add `narrow_backward_copy`? Or changing `narrow_tensor_symint` to `CompositeExplicitAutograd` will do.

kshitij12345 replied:
Regarding the error, the `narrow_backward(Tensor grad_output, SymInt[] input_sizes, int dim, Tensor start, SymInt length) -> Tensor(a)` is incorrect because `-> Tensor(a)` implies that you are returning one of the inputs but that is not the case as you allocate a new tensor `grad_input`. Changing the return from `-> Tensor(a)` to `-> Tensor` should mostly do the trick

isdanni committed:
Add Batch Rule for aten::narrow.Tensor (#99757)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/99757

Add batch support for `aten::narrow.Tensor`.

Similar logic to `narrow_copy`

Resolve issue: https://github.com/pytorch/pytorch/issues/99558

Test Plan: Please SEE contbuild & OSS CI.

Differential Revision: D45197155

fbshipit-source-id: 2e3a7be07940a072872446ae8b5492afa5098399

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1370,11 +1370,20 @@ Tensor narrow_symint(const Tensor& self, int64_t dim, SymInt start, SymInt lengt
 // This overload exists purely for XLA, because they wanted to pass in "symbolic"
 // start via Tensor.
-Tensor narrow_tensor_symint(const Tensor& self, int64_t dim, const Tensor& start, SymInt length) {
+Tensor narrow_tensor_symint(const Tensor& grad, IntArrayRef input_sizes, int64_t dim, const Tensor& start, SymInt length) {
```
I guess changing this function is a typo?

isdanni replied:
I updated `narrow_tensor_symint` in [aten/src/ATen/native/native_functions.yaml](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml) to `CompositeExplicitAutograd` according to [this review](https://github.com/pytorch/pytorch/pull/99757#discussion_r1180745927) so the function should be updated as well right?

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4010,7 +4010,14 @@
-    CompositeImplicitAutograd: narrow_tensor_symint
+    CompositeExplicitAutograd: narrow_tensor_symint
+
+- func: narrow_backward(Tensor grad_output, SymInt[] input_sizes, int dim, Tensor start, SymInt length) -> Tensor
```
Do you need this? Can't you use `narrow_scatter()` ?

isdanni replied:
> narrow_scatter

I checked and there is no `narrow_scatter` function like `slice_scatter`. So defining a `narrow_backward` should be the go-to way to add batching rules with `CompositeImplicitAutograd` like [this comment said](https://github.com/pytorch/pytorch/pull/99757#issuecomment-1526109834)?

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -3011,3 +3011,9 @@
   self: at::where(self[i] == scalars[i], grads[i] / 2, grads[i]).masked_fill_(self[i] < scalars[i], 0)
+
+- name: narrow.Tensor(Tensor(a) self, int dim, Tensor start, SymInt length) -> Tensor(a)
+  self: narrow_tensor_symint(grad, self.sym_sizes(), dim, start, length)
```
That should be the backward?

isdanni replied:
thx good catch! `grad.narrow_symint(grad, dim, start, length)`

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -3011,3 +3011,9 @@
+  self: narrow_tensor_symint(grad, self.sym_sizes(), dim, start, length)
+
+- name: narrow_backward(Tensor grad_output, SymInt[] input_sizes, int dim, Tensor start, SymInt length) -> Tensor
+  self: narrow_backward_symint(grad, self.sym_sizes(), dim, start, length)
```
That should be the double backward?

isdanni commented:
Closing following [dicussion](https://github.com/pytorch/pytorch/issues/99558#issuecomment-1531714799).
